### PR TITLE
Multidb support for SmartRedis

### PR DESCRIPTION
**Pull request by billschereriii**
Add multiDB support to SmartRedis using the ConfigOptions object. Update unit tests, examples, CI/CD scripts

API BREAK: The Client constructor for all languages is updated with a new signature. For langauges other than Python, the old signature remains available but deprecated. For Python, the order of the first two parameters was swapped in the old signature and it is still deprecated.

NOTE: This PR should not be merged until the SmartSim multiDB support PR is ready to merge as well.

### CODE REVIEW

File: conftest.py [hunk 1]
```diff
@@ -55,7 +55,11 @@
 
 @pytest.fixture
 def use_cluster():
-    return os.getenv('SMARTREDIS_TEST_CLUSTER', "").lower() == 'true'
+    if os.getenv('SR_SERVER_TYPE', "") == "":
```
Comment (amandarichardsonn) [inline conftest.py:58]: this just ignores if the env is not an empty string?

Comment (billschereriii) [reply conftest.py]: The entire fixture is gone now, so this is moot

File: utils/create_cluster/local_cluster.py [hunk 2]
```diff
@@ -80,17 +80,20 @@ def create_cluster(n_nodes, port):
...
-    if proc.returncode != 0:
-        raise SubprocessError("Cluster could not be created!")
+    if n_nodes > 1:
+        redis_cli = os.getenv('REDIS_INSTALL_PATH') + '/redis-cli'
+        cluster_str=' '
+        for i in range(n_nodes):
+            cluster_str += '127.0.0.1:' + str(port+i) + ' '
+        cmd = " ".join((redis_cli, "--cluster create", cluster_str, "--cluster-replicas 0"))
+        print(cmd)
+        proc = run([cmd], input="yes", encoding="utf-8", shell=True)
```
Comment (MattToast) [inline utils/create_cluster/local_cluster.py:90]: Tangential to this pr, but
```py
proc = run([cmd], input="yes", encoding="utf-8", shell=True)
         # ^^^^^ Little thrown off by the list of length 1 with str with spaces
         #       I'm guessing this is due to `shell=True` weirdness
         #       Can we change to `cmd.split()` so that each symbol is its own element
         #       in the list just to conform to the `subprocess.run` docs?
```

File: tests/python/test_errors.py [hunk 3]
```diff
@@ -32,90 +32,90 @@
...
+def test_bad_get_tensor(context):
+    c = Client(None, logger_name=context)
     with pytest.raises(RedisReplyError):
         c.get_tensor("not-a-key")
 
 
-def test_bad_get_dataset(use_cluster, context):
-    c = Client(None, use_cluster, logger_name=context)
+def test_bad_get_dataset(context):
+    c = Client(None, logger_name=context)
```
Comment (MattToast) [inline tests/python/test_errors.py:59]: Do we have any tests that still use the old `Client(cluser: bool, address: str, logger_name: str)` style constructor? If not, should we keep/add one until we formally remove that constructor?

**Review by MattToast — CHANGES_REQUESTED** [review]
I gave a quick first pass over the python/c/cpp side of things and everything seems to look great! I have a couple of small refactor requests, and I haven't looked into the tests in depth yet, but so far lgtm!!

**Review by ashao — CHANGES_REQUESTED** [review]
Some small changes requested but overall looks really good.

File: src/python/module/smartredis/client.py [hunk 4]
```diff
@@ -41,24 +42,66 @@
...
+        be removed in the next release of the SmartRedis client.
+
         :param cluster: True if connecting to a redis cluster, defaults to False
-        :type cluster: bool, optional
+        :type cluster: bool
+        :param address: Address of the database
+        :type address: str, optional
         :param logger_name: Identifier for the current client
         :type logger_name: str
         :raises RedisConnectionError: if connection initialization fails
```
[the remark this answers is not in the document]
Comment (billschereriii) [reply src/python/module/smartredis/client.py:107]: Add python-specific here

File: examples/parallel/cpp/smartredis_mnist.cpp [hunk 5]
```diff
@@ -104,8 +104,7 @@ int main(int argc, char* argv[]) {
     logger_name += std::to_string(rank);
 
     // Initialize a Client object
-    bool cluster_mode = true; // Set to false if not using a clustered database
-    SmartRedis::Client client(cluster_mode, logger_name);
+    SmartRedis::Client client(logger_name);
```
Comment (ashao) [inline examples/parallel/cpp/smartredis_mnist.cpp:107]: Especially if we want to encourage people to use the new route using ConfigOptions, we should update these examples. I'd actually be ok doing this in a separate PR as I think it might be a good educational task to do this and write a 'migration guide' for users since this is a fairly fundamental API break.

**Review by ashao — CHANGES_REQUESTED** [review]
Two small changes listed in the review. Also, if I could, could I request we change the `SR_SERVER_TYPE` to `SR_DB_TYPE` since we already use orchestrator and database as synonyms and I don't really want to introduce a third one?

File: doc/changelog.rst [hunk 6]
```diff
@@ -25,18 +25,10 @@ Detailed Notes
...
-- Many Fortran  routines were returning logical kind = c_bool which turns out not to be
-the same default kind of most Fortran compilers. These have now been refactored so that
-users need not import `iso_c_binding` in their own applications (PR340_)
+- Enable parallel build for the SmartRedis examples by moving utility Fortran code into a small static library (PR349_)
+- For the NVidia toolchain only: Replaces the assumed rank feature of F2018 used in the Fortran client with assumed shape arrays, making it possible to compile SmartRedis with the Nvidia toolchain. (PR346_)
+- Rework the build and test system to improve maintainability of the library. There have been several significant changes, including that Python and Fortran clients are no longer built by defaults and that there are Make variables that customize the build process. Please review the build documentation and ``make help`` to see all that has changed. (PR341_)
+- Many Fortran routines were returning logical kind = c_bool which turns out not to be the same default kind of most Fortran compilers. These have now been refactored so that users need not import `iso_c_binding` in their own applications (PR340_)
```
[the remark this answers is not in the document]
Comment (billschereriii) [reply doc/changelog.rst:69]: Fixed

File: src/python/module/smartredis/client.py [hunk 7]
```diff
@@ -40,19 +41,66 @@
...
+
+        For detailed information on the second signature, please refer
+        to the __deprecated_construction() method below.
+
+        :param a: The positional arguments supplied to this method; see above for
+                  valid options
+        :type a: tuple(any); see above for valid options
+        :param kw: Keyword arguments supplied to this method; see above for
+                   valid options
+        :type kw: dict {string, any}; see above for valid options
```
[the remark this answers is not in the document]
Comment (billschereriii) [reply src/python/module/smartredis/client.py:72]: done

File: doc/runtime.rst [hunk 8]
```diff
@@ -37,10 +37,21 @@ at three different addresses, each using port ``6379``:
...
     export SSDB="10.128.0.153:6379,10.128.0.154:6379,10.128.0.155:6379"
 
-The Python client relies on ``SSDB`` to determine database
-location.  However, the Python ``Client`` constructor also allows
-for the database location to be set as an input parameter. In
-this case, it sets SSDB from the input parameter.
+
+There are two types of Redis databases that can be used by the
+SmartRedis library. A ``Clustered`` database, such as the one in
+the previous example, Is replicated across multiple shards.
```
[the remark this answers is not in the document]
Comment (billschereriii) [reply doc/runtime.rst:43]: good eye

File: include/configoptions.h [hunk 9]
```diff
@@ -181,16 +202,22 @@ class ConfigOptions
...
-        void _set_log_context(const std::string& log_context) {
+        void _set_log_context(SRObject* log_context) {
             _log_context = log_context;
         }
 
+        /*!
+        *   \brief Clear a configuration option from the cache
+        *   \param option_name The name of the option to clear
+        */
+        void _clearOptionFromCache(const std::string& option_name);
```
[the remark this answers is not in the document]
Comment (billschereriii) [reply include/configoptions.h:219]: years and years of camelCase being my default -- I did this without thinking about it

File: src/cpp/rediscluster.cpp [hunk 10]
```diff
@@ -1018,7 +1019,12 @@ inline CommandReply RedisCluster::_run(const Command& cmd, std::string db_prefix
 // Connect to the cluster at the address and port
 inline void RedisCluster::_connect(SRAddress& db_address)
 {
+    std::string msg;
     for (int i = 1; i <= _connection_attempts; i++) {
+        msg = "Connection attempt " + std::to_string(i) + " of " +
+            std::to_string(_connection_attempts);
+        _cfgopts->_get_log_context()->log_data(LLDeveloper, msg);
```
[the remark this answers is not in the document]
Comment (billschereriii) [reply src/cpp/rediscluster.cpp:1026]: Good thought!

File: src/cpp/redisserver.cpp [hunk 11]
```diff
@@ -102,9 +103,19 @@ SRAddress RedisServer::_get_ssdb()
         address_choices.push_back(addr_spec);
     }
 
+    std::string msg = "Found " + std::to_string(address_choices.size()) + " addresses:";
+    _cfgopts->_get_log_context()->log_data(LLDeveloper, msg);
+    for (size_t ii = 0; ii < address_choices.size(); ii++) {
```
[the remark this answers is not in the document]
Comment (billschereriii) [reply src/cpp/redisserver.cpp:108]: Old versions of lint disliked single-letter variables and this was a common practive to make them happy. Not needed here, will change

File: src/python/module/smartredis/client.py [hunk 12]
```diff
@@ -62,7 +110,31 @@ def __init__(self, address=None, cluster=False, logger_name="default"):
         if "SSDB" not in os.environ:
             raise RedisConnectionError("Could not connect to database. $SSDB not set")
         try:
-            super().__init__(PyClient(cluster, logger_name))
+            return PyClient(cluster, logger_name)
+        except PybindRedisReplyError as e:
+            raise RedisConnectionError(str(e)) from None
+        except RuntimeError as e:
+            raise RedisConnectionError(str(e)) from None
```
[the remark this answers is not in the document]
Comment (billschereriii) [reply src/python/module/smartredis/client.py:118]: thanks for the pointer!

File: utils/create_cluster/local_cluster.py (hunk 2 above)
Comment (billschereriii) [reply utils/create_cluster/local_cluster.py:90]: We actually fixed this when we migrated to the new launch_redis.py script. This file is no longer in the project as of the test system rewrite

File: doc/changelog.rst [hunk 13]
```diff
@@ -1,6 +1,24 @@
 Changelog
 =========
 
+Changelog
```
[the remark this answers is not in the document]
Comment (billschereriii) [reply doc/changelog.rst:4]: Fixed as part of merge

File: doc/runtime.rst [hunk 14]
```diff
@@ -37,10 +37,21 @@ at three different addresses, each using port ``6379``:
...
+By way of comparison, a ``Standalone`` database only has a single
+shard that services all traffic; this is the form used when a
+colocated database is requested.
```
[the remark this answers is not in the document]
Comment (billschereriii) [reply doc/runtime.rst:46]: done

File: include/c_client.h [hunk 15]
```diff
@@ -42,6 +42,34 @@
...
 
+/*!
+*   \brief C-client simple constructor that uses default environment variables
+*          to locate configuration settings
+*   \param logger_name Identifier for the current client
+*   \param logger_name_length Length in characters of the logger_name string
+*   \param new_client Receives the new client
+*   \return Returns SRNoError on success or an error code on failure
+*/
+SRError CreateSimpleClient(
```
[the remark this answers is not in the document]
Comment (billschereriii) [reply include/c_client.h:53]: Done

File: include/configoptions.h [hunk 16]
```diff
@@ -98,22 +98,43 @@ class ConfigOptions
         */
         virtual ~ConfigOptions();
 
+        /*!
+        *   \brief Deep copy a ConfigOptions object
+        *   \returns The cloned object
+        *   \throw std::bad_alloc on allocation failure
+        */
+       ConfigOptions* clone();
```
[the remark this answers is not in the document]
Comment (billschereriii) [reply include/configoptions.h:106]: We call the deepcopy operation clone() in most places in the codebase, so I'd rather stick with precedent

File: include/logger.h [hunk 17]
```diff
@@ -31,6 +31,7 @@
 
 #ifdef __cplusplus // Skip C++ headers for C users
 #include <string>
+#include <iostream>
```
[the remark this answers is not in the document]
Comment (billschereriii) [reply include/logger.h:34]: removed

File: src/c/c_client.cpp [hunk 18]
```diff
@@ -67,8 +67,55 @@ auto c_client_api(T&& client_api_func, const char* name)
 #define MAKE_CLIENT_API(stuff)\
     c_client_api([&] { stuff }, __func__)()
 
+// Create a simple Client
+SRError CreateSimpleClient(
```
[the remark this answers is not in the document]
Comment (billschereriii) [reply src/c/c_client.cpp:71]: I did a global search/replace to change CreateSimpleClient to SimpleCreateClient() so this would have been included

File: tests/cpp/unit-tests/test_client.cpp [hunk 19]
```diff
@@ -770,6 +769,7 @@ SCENARIO("Test CONFIG SET on an unsupported command", "[Client]")
     log_data(context, LLDebug, "***End Client config get unsupported testing***");
 }
 
+#if TEST_REMOVAL
```
[the remark this answers is not in the document]
Comment (billschereriii) [reply tests/cpp/unit-tests/test_client.cpp:772]: debugging code; removed

File: src/python/module/smartredis/client.py [hunk 20]
```diff
@@ -40,19 +41,66 @@
...
+        still supported. Note that the order was swapped for first two
+        parameters in the second signature relative to previous releases
+        of SmartRedis; this was necessary to remove ambiguity. Support for
+        the second signature will be removed in a future version of the
+        SmartRedis library.
+
+            Client(config_options: ConfigOptions=None,
+                   logger_name: str="Default")
+            Client(cluster: bool, address: optional(str)=None,
+                   logger_name: str="Default") <= Deprecated!
```
[the remark this answers is not in the document]
Comment (billschereriii) [reply src/python/module/smartredis/client.py:59]: We can't until after we deprecate the old constructor in the python class, so I'll remove the definition for now

File: doc/changelog.rst (hunk 13 above)
[the remark this answers is not in the document]
Comment (billschereriii) [reply doc/changelog.rst:4]: Fixed

File: src/python/module/smartredis/client.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (billschereriii) [reply src/python/module/smartredis/client.py:107]: Added deprecation warning

**Comment by billschereriii** [comment]
> Two small changes listed in the review. Also, if I could, could I request we change the `SR_SERVER_TYPE` to `SR_DB_TYPE` since we already use orchestrator and database as synonyms and I don't really want to introduce a third one?

Renamed SR_SERVER_TYPE to SR_DB_TYPE. Good catch!

File: src/cpp/client.cpp [hunk 21]
```diff
@@ -28,32 +28,134 @@
...
+    _use_tensor_prefix = true;
+    _use_dataset_prefix = true;
+    _use_model_prefix = false;
+    _use_list_prefix = true;
+}
+
+
+// Simple Constructor
+Client::Client(const std::string& logger_name)
+    : SRObject(logger_name)
```
[the remark this answers is not in the document]
Comment (billschereriii) [reply src/cpp/client.cpp:66]: Done. C++11 added delegating constructors which makes this possible

File: tests/python/test_errors.py (hunk 3 above)
Comment (billschereriii) [reply tests/python/test_errors.py:59]: There is one in address.py that does!

File: src/cpp/client.cpp [hunk 22]
```diff
@@ -28,32 +28,134 @@
...
+
+    // Establish our server connection
+    _establish_server_connection();
+
+    // Initialize key prefixing
+    _get_prefix_settings();
+    _use_tensor_prefix = true;
+    _use_dataset_prefix = true;
+    _use_model_prefix = false;
+    _use_list_prefix = true;
```
[the remark this answers is not in the document]
Comment (billschereriii) [reply src/cpp/client.cpp:60]: Oversight, corrected

File: src/python/module/smartredis/client.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (billschereriii) [reply src/python/module/smartredis/client.py:59]: While we can't export this constructor publicly until we get rid of the deprecated Client constructor, we do use it with the new Client constructor in the case where the ConfigOptions object is `None`

File: examples/parallel/cpp/smartredis_mnist.cpp (hunk 5 above)
Comment (billschereriii) [reply examples/parallel/cpp/smartredis_mnist.cpp:107]: I agree both that this should be done and that it's a separate PR

File: tests/c/client_test_dataset_aggregation.c [hunk 23]
```diff
@@ -135,7 +135,7 @@ int main(int argc, char* argv[])
 
   // Initialize client
```
[the remark this answers is not in the document]
Comment (billschereriii) [reply tests/c/client_test_dataset_aggregation.c:136]: Will do!

**Review by ashao — APPROVED** [review]
Thanks for responding to the changes

**Review by MattToast — CHANGES_REQUESTED** [review]
One small question regrading env var prefix/suffixing implimentation, but otherwise LGTM!!

File: src/cpp/configoptions.cpp [hunk 24]
```diff
@@ -197,9 +215,9 @@ void ConfigOptions::_populate_options()
     );
 }
 
-// Apply a prefix to a option_name if the source is environment
-// variables and the prefix is nonempty
-std::string ConfigOptions::_prefixed(const std::string& option_name)
+// Apply a suffix to a option_name if the source is environment
+// variables and the suffix is nonempty
+std::string ConfigOptions::_suffixed(const std::string& option_name)
```
[the remark this answers is not in the document]
Comment (billschereriii) [reply src/cpp/configoptions.cpp:220]: Fixed

File: include/configoptions.h [hunk 25]
```diff
@@ -98,22 +99,43 @@ class ConfigOptions
...
         */
         static std::unique_ptr<ConfigOptions> create_from_environment(
-            const std::string& db_prefix);
+            const std::string& db_suffix);
+
+        /*!
+        *   \brief Instantiate ConfigOptions, getting selections from
+        *          environment variables. If \p db_suffix is non-empty,
+        *          then "{db_suffix}_" will be prepended to the name of
+        *          each environment variable that is read.
```
[the remark this answers is not in the document]
Comment (billschereriii) [reply include/configoptions.h:130]: fixed

File: include/configoptions.h [hunk 26]
```diff
@@ -98,22 +99,43 @@ class ConfigOptions
...
+
         /////////////////////////////////////////////////////////////
         // Factory construction methods
 
         /*!
         *   \brief Instantiate ConfigOptions, getting selections from
-        *          environment variables. If \p db_prefix is non-empty,
-        *          then "{db_prefix}_" will be prepended to the name of
+        *          environment variables. If \p db_suffix is non-empty,
+        *          then "{db_suffix}_" will be prepended to the name of
```
[the remark this answers is not in the document]
Comment (billschereriii) [reply include/configoptions.h:115]: fixed (here and many other places)

**Review by MattToast — APPROVED** [review]
LGTM! Thanks for the hard work on this one!!
